sycl/plugins/unified_runtime/ur/adapters/hip/queue.cpp

```diff
@@ -252,19 +255,44 @@ urQueueGetNativeHandle(ur_queue_handle_t hQueue, ur_queue_native_desc_t *,
 }
 
 /// Created a UR queue object from a HIP queue handle.
-/// TODO: Implement this.
-/// NOTE: The created UR object takes ownership of the native handle.
+/// NOTE: The created UR object doesn't takes ownership of the native handle.
 ///
 /// \param[in] hNativeQueue The native handle to create UR queue object from.
 /// \param[in] hContext is the UR context of the queue.
 /// \param[out] phQueue Set to the UR queue object created from native handle.
-/// \param pProperties->isNativeHandleOwned tells if SYCL RT should assume the
-/// ownership of
-///        the native handle, if it can.
-///
-/// \return UR_RESULT_ERROR_UNSUPPORTED_FEATURE
 UR_APIEXPORT ur_result_t UR_APICALL urQueueCreateWithNativeHandle(
-    ur_native_handle_t, ur_context_handle_t, ur_device_handle_t,
-    const ur_queue_native_properties_t *, ur_queue_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_native_handle_t hNativeQueue, ur_context_handle_t hContext,
+    ur_device_handle_t hDevice, const ur_queue_native_properties_t *pProperties,
+    ur_queue_handle_t *phQueue) {
+  (void)pProperties;
+  (void)hDevice;
+
+  unsigned int HIPFlags;
+  hipStream_t HIPStream = reinterpret_cast<hipStream_t>(hNativeQueue);
+
+  auto Return = UR_CHECK_ERROR(hipStreamGetFlags(HIPStream, &HIPFlags));
+
+  ur_queue_flags_t Flags = 0;
+  if (HIPFlags == hipStreamDefault)
+    Flags = UR_QUEUE_FLAG_USE_DEFAULT_STREAM;
+  else if (HIPFlags == hipStreamNonBlocking)
+    Flags = UR_QUEUE_FLAG_SYNC_WITH_DEFAULT_STREAM;
+  else
+    detail::ur::die("Unknown hip stream");
+
+  std::vector<hipStream_t> ComputeHIPStreams(1, HIPStream);
+  std::vector<hipStream_t> TransferHIPStreams(0);
+
+  // Create queue and set num_compute_streams to 1, as computeHIPStreams has
+  // valid stream
+  *phQueue = new ur_queue_handle_t_{std::move(ComputeHIPStreams),
+                                    std::move(TransferHIPStreams),
+                                    hContext,
+                                    hContext->getDevice(),
+                                    HIPFlags,
+                                    Flags,
+                                    /*backend_owns*/ false};
```

Comment:
we get the value indicating whether the backend owns the handle or not from isNativeHandleOwned from the properties. Shouldn't we use it instead of hardcoding to false?

```cpp
struct ur_queue_native_properties_t
Properties for for [urQueueCreateWithNativeHandle](https://oneapi-src.github.io/unified-runtime/core/api.html?highlight=ur_queue_native_properties_t#ur__api_8h_1a7a84f43f9fee0e0bf12ee9bce0d14a87).

Public Members

[ur_structure_type_t](https://oneapi-src.github.io/unified-runtime/core/api.html?highlight=ur_queue_native_properties_t#_CPPv419ur_structure_type_t) stype
[in] type of this structure, must be [UR_STRUCTURE_TYPE_QUEUE_NATIVE_PROPERTIES](https://oneapi-src.github.io/unified-runtime/core/api.html?highlight=ur_queue_native_properties_t#ur__api_8h_1aa3358252d6823fa2fa82b4862d3c957da9b3e2e81a606a0daf652f7feab2d359c)

void *pNext
[in,out][optional] pointer to extension-specific structure

bool isNativeHandleOwned
[in] Indicates UR owns the native handle or if it came from an interoperability operation in the application that asked to not transfer the ownership to the unified-runtime.
```

Reply:
It was decided in the CUDA backend spec for dpc++ that when creating with native objects ownership should not be passed to the SYCL object: https://github.com/KhronosGroup/SYCL-Docs/pull/420
We also decided HIP should have the same behavior, so I've simply copied what happens in UR CUDA, see:
https://github.com/intel/llvm/blob/b98e51ba336fb0dc2bce99cbe1d5f3243464f3d8/sycl/plugins/unified_runtime/ur/adapters/cuda/queue.cpp#L272

Now since UR can support different programming models than DPC++, I see that it may be desired to make the change you suggest. If you think it is important to make this change I can do it?
Otherwise it could be left as it is and changed if/and when another programming model outside DPC++ demands this flexibility of having the option to pass ownership away from the native object.

Thanks

Reply:
I've gone ahead and moved responsibility for deciding ownership to the DPC++ as suggested. Is this ok @jandres742 ?

Reply:
thanks @JackAKirk . I see your latest changes. Those are good. thanks.